### ISSUE TITLE
[DX] Remove direct use of `new Node()` and other direct instantiations of entity classes from the codebase 

### ISSUE THREAD

**Issue by argiepiano** [backdrop/backdrop-issues#6075]
When Backdrop partially incorporated D7 Entity API into core, and switched to using named classes for its core entities (that is, `Node` and `User` etc instead of `stdClass`), this allowed developers to extend those entity class (and their entity controllers) to incorporate additional functionalities and properties.  This can be done very simply and efficiently, by (1) implementing a `hook_entity_info_alter()` to change the entity class and the entity controller, and (2) extending the entity class and/or the entity controller class with the new functionality or properties.

Unfortunately, at the same time, a few places in Backdrop codebase instantiate those core entity classes by making direct calls, as in `$node = new Node();`. This effectively curtails the new possibilities created by the use of named classes as described above. So, if you are a developer who wants to transform/expand nodes in new ways, you are limited by those few direct instantiations (so, in those places, the resulting entity of calling `$node = new Node();` will use the core class, instead of the class you created to extend nodes).

The solution here is to replace those few `new Node()` and similar calls, with the intermediary API function `entity_create('node');`. This call will take care of instantiating the entity class as defined in the  entity_info_alter hook.

There are very few direct instantiations of core entity classes in the codebase:
- 7 for Node()
- 4 for User()
- 22 for File()
- 2 for Comment()
- 6 for TaxonomyTerm()

Many of those are in tests, which we can simply leave alone, as we want to use the core classes in tests.

**Comment by argiepiano** [comment]
Further, Drupal discourages instantiating classes directly. Instead their [coding standards](https://www.drupal.org/node/608152#instantiation) asks for the use of factory functions  - such as `entity_create()`.

**Comment by kiamlaluno** [comment]
[Object-oriented code / Instantiation](https://www.drupal.org/docs/develop/standards/php/object-oriented-code#instantiation)

**Comment by kiamlaluno** [comment]
That part of the coding standards is not followed by Drupal 10 itself. 😅

See the code in [book.module](https://api.drupal.org/api/drupal/core%21modules%21book%21book.module/10).

```php
$book_node = Node::load($node->book['bid']);
if (!$book_node->access()) {
  return;
}
```

(Do not get me wrong: I think the proposed change is the way to go. I find "funny" that Drupal has not fixed that after three major releases.)

**Comment by herbdool** [comment]
@argiepiano want to create a PR? I only see upside on this.

Hopefully not too many contrib modules instantiate the classes directly.

**Comment by laryn** [comment]
Sounds good to me, too.

**Comment by argiepiano** [comment]
Thanks. I'll submit a PR probably tonight.

**Comment by quicksketch** [comment]
> The solution here is to replace those few `new Node()` and similar calls, with the intermediary API function `entity_create('node');`. This call will take care of instantiating the entity class as defined in the entity_info_alter hook.

I don't think I've ever realized this would be something people would do. I'm not totally sure I'd recommend overriding the core classes with replacements; seems like may get us into compatibility problems if the sub-classes start creating methods that the core classes may eventually provide directly themselves (potentially with different purposes or signatures).

That said, I'm not opposed to this cleanup. I just worry that this might cause some headaches if a contrib module starts replacing some (or all) core entity classes.

**Comment by argiepiano** [comment]
PF https://github.com/backdrop/backdrop/pull/5036 provided.

**Comment by argiepiano** [comment]
>  I just worry that this might cause some headaches if a contrib module starts replacing some (or all) core entity classes

True, but ... the current code is half-way there. There are several instances of `entity_create('node' ,array())`, for example in the current code for `node_form()`. And that has not been cause for concern before.

**Comment by quicksketch** [comment]
Yeah, I had thought to use `entity_create()` when I need to handle multiple types of entities and when I knew I was working with a node or file I would just instantiate the object directly. I'm probably responsible for most of the current direct instantiations. I think the main reason I liked direct instantiation is that it made code completion a lot better in an IDE. When you call `entity_create()` the only type hint is that you're going to get `EntityInterface` back. That can (and probably should) be mitigated by a type hint on the assigned variable.

i.e. changes from `$node = new Node(array())` should be changed to:
```
/* @var Node $node */
$node = entity_create('node', array());
``` 

That keeps the type-specific type hinting in place. Though it might not be accurate if the entity type info gets altered. But again, I would expect that to be a high-risk choice to swap out an entire entity's class. But that doesn't mean a module like Entity Plus wouldn't do it, and then we could end up with large swaths of contrib built on a concept that would make core development harder in the future.

**Comment by argiepiano** [comment]
@quicksketch thanks. I'm a bit confused by the last comment - are you saying you are not supportive of using `entity_create()` and that we should revert that to direct instantiation? Can you please clarify?

**Comment by argiepiano** [comment]
BTW @quicksketch I tried what you suggested: 

```php
/* @var Node $node */
$node = entity_create('node', array());
```

It doesn't work in VS Code. `entity_create()` is already returning an object of class `EntityInterface`, so VS Code ignores the inline comment above the assignment, and instead uses `EntityInterface` to type hint.

I don't know if this might work in other IDE environments.

**Comment by indigoxela** [comment]
The code change looks good to me, and all tests passing (as expected). 👍 

The benefit of the change would be consistency. Not sure, how common it is, that contrib (custom) overrides the whole class, but extending should be possible - and that's where entity_create() brings consistency (IMO).

However, I belief, there's some misunderstanding re the type hint (special) comment.

I took a look with PHPStan and the type hint is indeed beneficial. Examples:

```
// PHPStan is able to determine the type of $node is "Node".
$node = new Node(array('title' => 'this is a test'));
```
But...
```
// PHPStan is not able to determine the correct type, assumes "EntityInterface".
$node = entity_create('node', array('title' => 'another test'));
```

But there's a solution for that:
```
  // PHPStan is able to get the correct type "Node".
  /** @var Node $node */
  $node = entity_create('node', array('title' => 'another test'));
```
Note that the comment has to be in this exact form, starting with `/**` - two asterisks.

See also: https://phpstan.org/writing-php-code/phpdocs-basics#inline-%40var

I'm not sure, what VS Code uses under the hood, but can you try (locally) with the correct comment syntax (with two asterisks), to see if that also works with VS Code @argiepiano ?

**Comment by herbdool** [comment]
This does work in VSCode (when using the PHP whatever extensions):

```
/** @var Node $node */
  $node = entity_create('node', array('title' => 'another test'));
```

**Comment by quicksketch** [comment]
I created a separate issue for the type hinting inconsistency at https://github.com/backdrop/backdrop-issues/issues/6876

**Comment by indigoxela** [comment]
> This does work in VSCode...

That's good news. Existing usage of the type hint comments get fixed in the spin-off issue, but it also means, this PR needs a little work - when switching from direct use to entity_create() we need the type hints.

@argiepiano I'm aware that your day job's currently quite demanding and it may take a little while until you're able to do the update - we'll wait.
Unless it's super-urgent - in that case someone could take over here. But I actually don't assume, that's the case.

**Comment by argiepiano** [comment]
Hi @indigoxela. I've added type hinting to all `entity_create()` instances in core. Ready for review. I had to change the name of a few variables in some test files, as the functions used the same name to provide an array of parameters to entity create, as in 

```php
/** @var File $file */
$file = entity_create('file', $file);
```

where the second parameter was an array to pass to `entity_create()`. This created some linting in VS Code, as the second parameter was an array.

**Comment by argiepiano** [comment]
Hmm... it appears that phpcs doesn't really like `/** @var Node $node */  etc.

What to do??

**Comment by indigoxela** [comment]
> it appears that phpcs doesn't really like /** @var Node $node */ etc...

Our phpcs ruleset is just fine with this sort of comment (`/** @var Node $node */`). Whatever fails with the action - it's something else. I'd suggest to just ignore it for now. My suspicion is, that too many files changed for this action to handle.

@argiepiano many thanks for this quick turnaround. 🙏

**Comment by argiepiano** [comment]
Thanks, @indigoxela. I have addressed the comment.

**Comment by indigoxela** [comment]
> I have addressed the comment.

Many thanks for complementing all the type hint comments. This should make working with our code easier. 👍 

This is ready to go in now, RTBC.

**Comment by quicksketch** [comment]
I think about this issue often still and although I am still very hesitant, I think this is probably the right thing to do. We discussed it a few more times in dev meetings and while others (such as @laryn and @jenlampton) share my concern about contrib overriding core classes, I don't think that should be enforced by a code-style coincidence. The API we inherited from Drupal 7 provides the ability to override the storage classes, as is literally documented in [Backdrop's hook_entity_info_alter()](https://docs.backdropcms.org/api/backdrop/core%21modules%21entity%21entity.api.php/function/hook_entity_info_alter/1), so it is even more strange that the storage class can be overridden but the entity class cannot.

So unless anyone has strong objections beyond "contrib could do something really weird with this" (my primary concern), I think I'm ready to concede and merge this change.

**Comment by laryn** [comment]
No strong objections.

**Comment by quicksketch** [comment]
Thanks folks, after consideration and discussion over months, I've merged https://github.com/backdrop/backdrop/pull/5036. A big thank you to @argiepiano for your patience in letting me work through the implications of this change.

https://github.com/backdrop/backdrop/commit/859d636f52e16f18f02e61025543412627bd4160 by @argiepiano, @avpaderno, @herbdool, @bugfolder, @quicksketch, @indigoxela, and @laryn.
